build/parseSpec.c

```diff
@@ -467,13 +514,21 @@ int readLine(rpmSpec spec, int strip)
 	goto retry;
     }
 
-    if (match != -1) {
+    if (isIf) {
 	rl = xmalloc(sizeof(*rl));
 	rl->reading = spec->readStack->reading && match;
 	rl->next = spec->readStack;
 	rl->lineNum = ofi->lineNum;
+	rl->readable = (!rl->reading) && (spec->readStack->reading);
+	rl->elifEnabled = 1;
 	spec->readStack = rl;
 	spec->line[0] = '\0';
+    } else if (isElif) {
+	spec->readStack->reading = match && spec->readStack->readable;
+	if (spec->readStack->reading)
+	    spec->readStack->readable = 0;
+	spec->line[0] = '\0';
+	match = -1;
```

Comment:
Sorry but this all still seems like something duct-taped to the side to me. Like I said earlier in private: now that you know what's involved, please rethink the approach from scratch - what you really need tracked and where, and how to best achieve these things in a clear and concise manner. One can almost always do a better job the second time around.

There would seem to be a fair amount of overlap in these various tracking variables, some local, some wider, that could be eliminated / combined. For example, isn't elifEnabled practically the same thing as checking whether ->next non-NULL? Also makes me wonder if it'd fit in more naturally if you just treated %elif as a new %if inside an %else, which is what it ultimately is. %elif is just syntactic sugar afterall.

Reply:
Thank you for the comment. 

> For example, isn't elifEnabled practically the same thing as checking 
> whether ->next non-NULL?

No. Because `elifEnabled = 1` if some `%if` condition starts and does not reach `%else` or `%endif`. `->next` is non-NULL if some `%if` starts and does not reach `%endif`. And from all other `->` variables you can not decide whether in the current `%if` `%else` was reached or not. Yes the meaning and name of the variable can be different, but there must be some variable containing this information.

> Also makes me wonder if it'd fit in more naturally if you just treated 
> %elif as a new %if inside an %else, which is what it ultimately is. 
> %elif is just syntactic sugar after all.

This alternative approach does not optimize variables, or complexity of the program:  
Programming `%elif` like a "special `%else` `%if` " will lead into:
- `->elifEnabled` must stay in the code with this approach too. (Of course no metter how it is implemented, there is need for testing whether all %elif's are on their propper places). 
- even if in this approach `->reading` can be omitted, there must be some new 
variable counting how nested the `%elif` is. It is because after `%else` resp. `%endif`, function `readLine`, must close the main `%if`, and all other "auxiliary %else %if's" that emulate `%elifs`.
Thus it will result into code that is not better than the current one.

Current approach corresponds to defining %elif straightforwardly according to its syntax and the alternative approach corresponds to defining %elif using %if inside an %else. Both are correct definitions, but for me the current looks more readable and natural.

Reply:
I wasn't asking for detailed explanation against my remarks, just  pointing (partial) overlap and throwing possible ideas to explore. Like said: knowing what's involved and the surrounding, rethink from scratch and see what comes up, you can almost always do better the second or even third time around. You mentioned corner cases that this doesn't take into account - take them into account from the beginning the in the next round. I'm absolutely positive that if-elif-else-endif branching can be expressed and tracked in a more coherent manner than we have here. As discussed in private this ugly code to begin with, take the opportunity to improve the foundation instead of duct-taping over it. Sometimes there's no choice but to duct-tape due to time or API constraints but neither is an issue here.

Reply:
> I wasn't asking for detailed explanation against my remarks, just pointing (partial) overlap and throwing possible ideas to explore. 

I think that if you had a concrete remark and I am not planning to involve it in next version of PR, then it is fair to explain why. And usually it is better to explain it precisely.

> Like said: knowing what's involved and the surrounding code, rethink from scratch and see what comes up, you can almost  always do better the second or even third time around. 

I rethink this code many times and for me it results in the refactorization patches before the "elif" patch. 

> You mentioned corner cases that this doesn't take into account - take them into account from the beginning the in the next round.

Yes I mentioned those corner cases. Of course I planned to solve them - but in following PR. There is no gain from solving it in this PR.

> I'm absolutely positive that if-elif-else-endif branching can be expressed and tracked in a more coherent manner than we have here. As discussed in private this ugly code to begin with, take the opportunity to improve the foundation instead of duct-taping over it. Sometimes there's no choice but to duct-tape due to time or API  constraints but neither is an issue here.

If you have some concrete issues or improvements of the PR, please send me them. But I am happy with this code.

Reply:
Well, I wouldn't be asking for you to rethink over and over again if *I* was happy with the code.
Which is why I've been asking for you to rethink at a more fundamental level, fundamental meaning something deeper than moving around a few lines of existing code.

So we have a bit of a stalemate here: if you can't see anything wrong with this code then clearly I'm not going to get anything better by asking for a rethink.